code/datums/components/barber.dm

```diff
@@ -448,6 +420,154 @@ TYPEINFO(/datum/component/barber)
 
 
 
+/datum/component/barber/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "BarberingMenu")
+		ui.open()
+
+/datum/component/barber/ui_data(mob/user)
+	if (isnull(src.new_AH))
+		src.new_AH = new /datum/appearanceHolder/()
+		src.new_AH.CopyOther(src.barbee.bioHolder.mobAppearance)
+
+	if (isnull(src.preview))
+		var/preview_id = src.barber.name + "_" + src.barbee.name + "_" + (istype(src, /datum/component/barber/shave) ? "shave" : "cut") // To avoid mixing up preview IDs, we gotta be *really* specific
+		src.preview = new /datum/character_preview(src.barbee.client, "barber", preview_id)
+		src.preview.add_background("#242424")
+		src.preview.mirror_appearance(barbee)
+		src.preview.update_appearance(src.barbee.bioHolder.mobAppearance, direction=SOUTH, name=src.barbee.name)
+
+	var/list/current_hair_style = list("bottom" = new_AH.customization_first.name, "middle" = new_AH.customization_second.name, "top" = new_AH.customization_third.name)
+	. = list("preview" = src.preview.preview_id, "selected_hair_portion" = hair_portion, "current_hair_style" = current_hair_style)
+
+/datum/component/barber/ui_static_data(mob/user)
+	var/list/all_hairs = list()
+	var/all_styles = null
+
+	if (istype(src, /datum/component/barber/shave))
+		all_styles = concrete_typesof(/datum/customization_style/beard) + concrete_typesof(/datum/customization_style/moustache) + concrete_typesof(/datum/customization_style/sideburns) + concrete_typesof(/datum/customization_style/eyebrows)
+	else
+		all_styles = concrete_typesof(/datum/customization_style/hair)
+
+	// just so we get a special icon sprite for no hair
+	all_hairs += list("None" = list("hair_id" = "none", "hair_icon" = "data:image/png;base64," + icon2base64(icon('icons/map-editing/landmarks.dmi', "x", SOUTH))))
+
+	for (var/datum/customization_style/styles as anything in all_styles)
+		var/datum/customization_style/style = new styles
+		var/hair_icon = "data:image/png;base64," + icon2base64(icon('icons/mob/human_hair.dmi', style.id, SOUTH, 1)) // yeah, sure, i'll keep it white. the user can preview the hair style anyway.
+		all_hairs += list(style.name = list("hair_id" = style.id, "hair_icon" = hair_icon))
+
+	. = list("available_styles" = all_hairs)
+
+/datum/component/barber/ui_act(var/action, var/params)
+	. = ..()
+	if (.)
+		return
+
+	switch(action)
+
+		if("change_hair_portion")
+			src.hair_portion = params["new_portion"]
+			return TRUE
+
+		if("do_hair")
+			if (isnull(params["style_id"])) // It means we are making a wig
+				if (istype(src, /datum/component/barber/shave))
+					actions.start_and_wait(new/datum/action/bar/barber/shave(src.barbee, src.barber, get_barbery_conditions(src.barbee, src.barber), null, ALL_HAIR), src.barber)
```

Comment:
Do `start_and_wait` actions do delay stuff? If so, probably need to check continued existence/validity of the various things we're acting on below?